src/main/java/com/sourcegraph/cody/autocomplete/CodyEditorFactoryListener.kt

```diff
@@ -0,0 +1,183 @@
+ * and document edits.
+ */
+class CodyEditorFactoryListener : EditorFactoryListener {
+  private var selectionListener = CodySelectionListener()
+  private var caretListener: CaretListener = CodyCaretListener()
+
+  override fun editorCreated(event: EditorFactoryEvent) {
+    if (!isCodyEnabled()) {
+      return
+    }
+    val editor = event.editor
+    informAgentAboutEditorChange(editor, skipCodebaseOnFileOpened = false)
+    val project = editor.project
+    if (project == null || project.isDisposed) {
+      return
+    }
+    val disposable = Disposer.newDisposable("CodyEditorFactoryListener")
+    EditorUtil.disposeWithEditor(editor, disposable)
+    editor.caretModel.addCaretListener(caretListener, disposable)
+    editor.selectionModel.addSelectionListener(selectionListener, disposable)
+    editor.document.addDocumentListener(CodyDocumentListener(editor), disposable)
+  }
+
+  private class CodyCaretListener : CaretListener {
+    override fun caretPositionChanged(e: CaretEvent) {
+      if (!isCodyEnabled()) {
+        return
+      }
+      val commandName = CommandProcessor.getInstance().currentCommandName
+      if (commandName == VIM_EXIT_INSERT_MODE_ACTION) {
+        return
+      }
+      informAgentAboutEditorChange(e.editor)
+      val suggestions = instance
+      val editor = e.editor
+      if (isEditorValidForAutocomplete(editor) && isSelectedEditor(editor)) {
+        suggestions.clearAutocompleteSuggestions(e.editor)
+        if (isImplicitAutocompleteEnabledForEditor(editor))
+            suggestions.triggerAutocomplete(
+                e.editor, e.editor.caretModel.offset, InlineCompletionTriggerKind.AUTOMATIC)
+      }
+    }
+  }
+
+  private class CodySelectionListener : SelectionListener {
+    override fun selectionChanged(e: SelectionEvent) {
+      if (!isCodyEnabled()) {
+        return
+      }
+      val editor = e.editor
+      informAgentAboutEditorChange(editor)
+      if (isEditorValidForAutocomplete(editor) && isCodyEnabled() && isSelectedEditor(editor))
+          instance.clearAutocompleteSuggestions(editor)
+    }
+  }
+
+  private class CodyDocumentListener(private val editor: Editor) : BulkAwareDocumentListener {
+    override fun documentChangedNonBulk(event: DocumentEvent) {
+      if (!isSelectedEditor(editor)) {
+        return
+      }
+      val completions = instance
+      completions.clearAutocompleteSuggestions(editor)
+      if (isImplicitAutocompleteEnabledForEditor(editor) &&
+          isEditorValidForAutocomplete(editor) &&
+          !CommandProcessor.getInstance().isUndoTransparentActionInProgress) {
+        informAgentAboutEditorChange(editor)
+        val changeOffset = event.offset + event.newLength
+        if (editor.caretModel.offset == changeOffset) {
+          completions.triggerAutocomplete(
+              editor, changeOffset, InlineCompletionTriggerKind.AUTOMATIC)
+        }
+      }
+    }
+  }
+
+  companion object {
```

Comment:
I'm getting this warning from IDE. Maybe it makes sense to create separate class for this part 😅 
<img width="925" alt="Screenshot 2023-10-26 at 15 02 18" src="https://github.com/sourcegraph/jetbrains/assets/9321940/aefa15c8-bc3e-425e-9a7c-d95e997e001b">

Reply:
Nice catch! How do you invoked this analysis? I'm trying to reproduce this locally, just for curiosity.